File: backend/internal/service/inference.py
from __future__ import annotations

import time
from uuid import UUID, uuid4

from dao.repositories.inference import InferenceRepository
from dao.repositories.worker import WorkerRepository
from pkg.apis.inference.models import InferenceRequest, ProtectedInferenceResponse
from pkg.runtime.status import InferenceStatus
from pkg.security.enforcement import EnforcementAction, Enforcer
from pkg.serving.router import ServingRouter

from internal.service.security import SecurityService
from internal.telemetry.logging import get_logger
from internal.telemetry.metrics import SpiderMetrics

logger = get_logger("spider-inference")


class InferenceService:
    def __init__(
        self,
        security: SecurityService,
        enforcer: Enforcer,
        router: ServingRouter,
        inferences: InferenceRepository,
        workers: WorkerRepository,
        metrics: SpiderMetrics,
    ) -> None:
        self.security = security
        self.enforcer = enforcer
        self.router = router
        self.inferences = inferences
        self.workers = workers
        self.metrics = metrics

    async def infer(
        self,
        request: InferenceRequest,
        *,
        user_id: UUID | None = None,
    ) -> ProtectedInferenceResponse:
        request_id = uuid4()
        started = time.perf_counter()

        security_result, scan = await self.security.inspect(
            request.prompt,
            request_id=request_id,
            source="inference",
            user_id=user_id,
            model=request.model,
            persist=True,
        )
        security_overhead_ms = security_result.total_latency_ms
        action = self.enforcer.resolve(security_result.decision)

        if action is not EnforcementAction.FORWARD:
            status = (
                InferenceStatus.REVIEW.value
                if action is EnforcementAction.HOLD
                else InferenceStatus.BLOCKED.value
            )
            e2e = (time.perf_counter() - started) * 1000.0
            record = await self.inferences.create(
                request_id=request_id,
                model=request.model,
                status=status,
                decision=security_result.decision.value,
                scan_id=scan.id if scan else None,
                user_id=user_id,
                worker_id=None,
                end_to_end_latency_ms=e2e,
                security_overhead_ms=security_overhead_ms,
                inference_latency_ms=None,
                output_preview=None,
            )
            await self.inferences.add_event(
                record.id,
                "blocked" if status == InferenceStatus.BLOCKED.value else "held",
                {"action": action.value},
            )
            self.metrics.observe_inference(
                status=status,
                e2e_latency_ms=e2e,
                security_overhead_ms=security_overhead_ms,
            )
            logger.info(
                "inference_not_forwarded",
                request_id=str(request_id),
                decision=security_result.decision.value,
                action=action.value,
            )
            return ProtectedInferenceResponse(
                request_id=request_id,
                status=status,
                decision=security_result.decision,
                model=request.model,
                output=None,
                security=security_result,
                security_overhead_ms=security_overhead_ms,
                inference_latency_ms=None,
                end_to_end_latency_ms=e2e,
            )

        worker_inventory = await self.workers.list_resources()
        llm_started = time.perf_counter()
        llm_response, selected = await self.router.route(request, workers=worker_inventory)
        inference_latency_ms = (time.perf_counter() - llm_started) * 1000.0
        e2e = (time.perf_counter() - started) * 1000.0
        worker_id = selected.worker_id if selected else None
        preview = (llm_response.output or "")[:512] or None

        record = await self.inferences.create(
            request_id=request_id,
            model=request.model,
            status=InferenceStatus.COMPLETED.value,
            decision=security_result.decision.value,
            scan_id=scan.id if scan else None,
            user_id=user_id,
            worker_id=worker_id,
            end_to_end_latency_ms=e2e,
            security_overhead_ms=security_overhead_ms,
            inference_latency_ms=inference_latency_ms,
            output_preview=preview,
        )
        await self.inferences.add_event(
            record.id,
            "completed",
            {"provider": getattr(self.router.provider, "name", "unknown")},
        )
        self.metrics.observe_inference(
            status=InferenceStatus.COMPLETED.value,
            e2e_latency_ms=e2e,
            security_overhead_ms=security_overhead_ms,
        )
        logger.info(
            "inference_completed",
            request_id=str(request_id),
            decision=security_result.decision.value,
            worker_id=worker_id,
            model=request.model,
            latency_ms=e2e,
        )
        return ProtectedInferenceResponse(
            request_id=request_id,
            status=InferenceStatus.COMPLETED.value,
            decision=security_result.decision,
            model=request.model,
            output=llm_response.output,
            security=security_result,
            security_overhead_ms=security_overhead_ms,
            inference_latency_ms=inference_latency_ms,
            end_to_end_latency_ms=e2e,
            worker_id=worker_id,
        )

File: backend/internal/service/inference.go
package service

import (
	"context"
	"log/slog"
	"time"

	"github.com/google/uuid"
	"github.com/spider/spider/dao/store"
	"github.com/spider/spider/internal/telemetry"
	"github.com/spider/spider/pkg/apis"
	"github.com/spider/spider/pkg/runtime"
	"github.com/spider/spider/pkg/security/enforcement"
	"github.com/spider/spider/pkg/serving/router"
)

type InferenceService struct {
	Security   *SecurityService
	Enforcer   *enforcement.Enforcer
	Router     *router.ServingRouter
	Inferences InferenceStore
	Workers    WorkerStore
	Metrics    *telemetry.Metrics
}

func (s *InferenceService) Infer(ctx context.Context, request apis.InferenceRequest, userID *uuid.UUID) (apis.ProtectedInferenceResponse, error) {
	requestID := uuid.New()
	started := time.Now()

	result, scan, err := s.Security.Inspect(ctx, request.Prompt, InspectOptions{
		RequestID: requestID.String(),
		Source:    strPtr("inference"),
		UserID:    userID,
		Model:     &request.Model,
		Persist:   true,
	})
	if err != nil {
		return apis.ProtectedInferenceResponse{}, err
	}

	securityOverhead := result.TotalLatencyMs
	action := s.Enforcer.Resolve(result.Decision)

	if action != enforcement.ActionForward {
		status := runtime.InferenceStatusBlocked
		if action == enforcement.ActionHold {
			status = runtime.InferenceStatusReview
		}
		e2e := float64(time.Since(started).Microseconds()) / 1000.0
		var scanID *uuid.UUID
		if scan != nil {
			scanID = &scan.ID
		}
		rec, err := s.Inferences.Create(ctx, store.CreateInferenceParams{
			RequestID: requestID, Model: request.Model, Status: status,
			Decision: string(result.Decision), ScanID: scanID, UserID: userID,
			EndToEndLatencyMs: e2e, SecurityOverheadMs: securityOverhead,
		})
		if err != nil {
			return apis.ProtectedInferenceResponse{}, err
		}
		eventType := "blocked"
		if status == runtime.InferenceStatusReview {
			eventType = "held"
		}
		_ = s.Inferences.AddEvent(ctx, rec.ID, eventType, map[string]interface{}{"action": string(action)})
		s.Metrics.ObserveInference(status, e2e, securityOverhead)
		slog.Info("inference_not_forwarded", "request_id", requestID.String(), "decision", result.Decision, "action", action)
		return apis.ProtectedInferenceResponse{
			RequestID: requestID.String(), Status: status, Decision: result.Decision,
			Model: request.Model, Security: result, SecurityOverheadMs: securityOverhead,
			EndToEndLatencyMs: e2e,
		}, nil
	}

	inventory, err := s.Workers.ListResources(ctx)
	if err != nil {
		return apis.ProtectedInferenceResponse{}, err
	}
	llmStarted := time.Now()
	llmResponse, selected, err := s.Router.Route(ctx, request, inventory)
	if err != nil {
		return apis.ProtectedInferenceResponse{}, err
	}
	inferenceLatency := float64(time.Since(llmStarted).Microseconds()) / 1000.0
	e2e := float64(time.Since(started).Microseconds()) / 1000.0

	var workerID *string
	if selected != nil {
		workerID = &selected.WorkerID
	}
	var scanID *uuid.UUID
	if scan != nil {
		scanID = &scan.ID
	}
	var preview *string
	if llmResponse.Output != nil {
		p := *llmResponse.Output
		if len(p) > 512 {
			p = p[:512]
		}
		preview = &p
	}

	rec, err := s.Inferences.Create(ctx, store.CreateInferenceParams{
		RequestID: requestID, Model: request.Model, Status: runtime.InferenceStatusCompleted,
		Decision: string(result.Decision), ScanID: scanID, UserID: userID, WorkerID: workerID,
		EndToEndLatencyMs: e2e, SecurityOverheadMs: securityOverhead,
		InferenceLatencyMs: &inferenceLatency, OutputPreview: preview,
	})
	if err != nil {
		return apis.ProtectedInferenceResponse{}, err
	}
	_ = s.Inferences.AddEvent(ctx, rec.ID, "completed", map[string]interface{}{"provider": "mock"})
	s.Metrics.ObserveInference(runtime.InferenceStatusCompleted, e2e, securityOverhead)
	slog.Info("inference_completed", "request_id", requestID.String(), "decision", result.Decision, "worker_id", workerID, "latency_ms", e2e)

	return apis.ProtectedInferenceResponse{
		RequestID: requestID.String(), Status: runtime.InferenceStatusCompleted,
		Decision: result.Decision, Model: request.Model, Output: llmResponse.Output,
		Security: result, SecurityOverheadMs: securityOverhead,
		InferenceLatencyMs: &inferenceLatency, EndToEndLatencyMs: e2e, WorkerID: workerID,
	}, nil
}
